ovssubnet/common.go

```diff
@@ -123,13 +124,36 @@ func (oc *OvsController) StartMaster(sync bool, containerNetwork string, contain
 	if err != nil {
 		return err
 	}
-	err = oc.ServeExistingNodes()
+
+	// SDN needs to fetch current items in etcd and watch for any new
+	// changes for resources like nodes, subnets, namespaces, services
+	// and netnamespaces.
+	//
+	// To avoid any potential race conditions during this process, these steps are followed:
+	// 1. Initiator(master/node): Watch for a resource as an async op, lets say WatchProcess
+	// 2. Initiator: Ensure watch for the resource is ready
+	//    This is needed as step-1 is an asynchronous operation
+	// 3. WatchProcess: When ready for watching, send ready signal to initiator
+	// 4. WatchProcess: Collect new changes in the queue but wait for initiator
+	//    to indicate which version to start from
+	// 5. Initiator: Get existing items with their latest version for the resource
+	// 6. Initiator: Send version from step-5 to WatchProcess
+	// 7. WatchProcess: Ignore any items with version <= start version got from initiator on step-6
+	// 8. WatchProcess: Handle new changes
+
+	ready := make(chan bool)
+	start := make(chan string)
+	go oc.watchNodes(ready, start)
+	waitForWatchReadiness(ready, "Node")
+
+	version, err := oc.serveExistingNodes()
 	if err != nil {
-		log.Warningf("Error initializing existing nodes: %v", err)
-		// no worry, we can still keep watching it.
+		return err
 	}
+	start <- version
+
```

Comment:
This pattern occurs 5 times. I think it should be possible to turn it into a function? (Maybe it would be awkward because of types.)